### PR TITLE
SPORTSHUB-64 Sendgrid Integration

### PR DESCRIPTION
**Pull request by owya490**

### CODE REVIEW

File: frontend/components/navbar/ProfilePic.tsx [hunk 1]
```diff
@@ -86,7 +87,8 @@ export default function ProfilePic() {
         <button
           className="px-4 py-2 rounded-lg mx-3 max-h-[40px] font-semibold hidden lg:block whitespace-nowrap  bg-highlight-yellow text-white hover:bg-white hover:text-highlight-yellow border-2 border-highlight-yellow"
           onClick={() => {
-            router.push("/event/create");
+            sendEmailOnCreateEvent("tHa1iwYdcNeB29RrBBhL", "Public");
+            // router.push("/event/create");
```
Comment (ReggieStar11) [inline frontend/components/navbar/ProfilePic.tsx:91]: delete this, seems unneccesary

File: functions/lib/sendgrid/commons.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from lib.constants import db
+from lib.logging import Logger
+
+CREATE_EVENT_EMAIL_TEMPLATE_ID = "d-63eb299010344bcf951af2537c42a410"
```
Comment (ReggieStar11) [inline functions/lib/sendgrid/commons.py:4]: is it possible to put this somewhere hidden, i think its bad to have this out

File: functions/lib/sendgrid/create_event.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  eventId: str
+  visibility: str
+
+  def __post_init__(self):
+    if not isinstance(self.eventId, str):
+      raise ValueError("Event Id must be provided as a string.")
+    if not isinstance(self.visibility, str):
+      raise ValueError("Visibility must be provided as a string.")
+
+#cors=options.CorsOptions(cors_origins=["localhost", "www.sportshub.net.au", "*"], cors_methods=["post"])
```
Comment (ReggieStar11) [inline functions/lib/sendgrid/create_event.py:31]: delete this

File: functions/lib/sendgrid/purchase_event.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  orderId: str
+
+  def __post_init__(self):
+    if not isinstance(self.eventId, str):
+      raise ValueError("Event Id must be provided as a string.")
+    if not isinstance(self.visibility, str):
+      raise ValueError("Visibility must be provided as a string.")
+
+
+# @https_fn.on_call(cors=options.CorsOptions(cors_origins=["localhost", "www.sportshub.net.au", "*"], cors_methods=["post"]), region="australia-southeast1")
```
Comment (ReggieStar11) [inline functions/lib/sendgrid/purchase_event.py:36]: do we need this

File: functions/lib/sendgrid/purchase_event.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    end_date_string =  end_date.ToDatetime().strftime("%m/%d/%Y, %H:%M")
+    logger.info(start_date_string)
+    logger.info(end_date_string)
+
+    message.dynamic_template_data = {
+      "first_name": request_data.first_name,
+      "event_name": event_data.get("name"),
+      "order_id": request_data.orderId,
+      "location": event_data.get("location"),
+      "quantity_bought": len(order_data.get("tickets")),
```
Comment (ReggieStar11) [inline functions/lib/sendgrid/purchase_event.py:79]: maybe change the name of "tickets" could be confusing

**Review by ReggieStar11 — CHANGES_REQUESTED** [review]

File: functions/lib/sendgrid/commons.py (hunk 2 above)
Comment (owya490) [reply functions/lib/sendgrid/commons.py:4]: email templates should be public domain knowledge, i don't think these need to be kept private.. unless sendgrid advises somwhere?

File: functions/lib/sendgrid/purchase_event.py (hunk 5 above)
Comment (owya490) [reply functions/lib/sendgrid/purchase_event.py:79]: its the list of TicketIds. What do you mean change the name of "tickets"?

i use the length of Tickets to determine how many tickets are bought

File: functions/lib/sendgrid/purchase_event.py (hunk 4 above)
Comment (owya490) [reply functions/lib/sendgrid/purchase_event.py:36]: nice catch, deleted!

File: functions/lib/sendgrid/create_event.py (hunk 3 above)
Comment (owya490) [reply functions/lib/sendgrid/create_event.py:31]: like wise! deleted!

File: frontend/services/src/sendgrid/sendgridService.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+  const sendEmailFunction = getFirebaseFunctionByName(FIREBASE_FUNCTIONS_SEND_EMAIL_ON_CREATE_EVENT);
+  return sendEmailFunction(content)
+    .then((result) => {
+      const data = JSON.parse(result.data as string) as SendgridSendEmailOnCreateEvent;
+      return data.status;
+    })
+    .catch((error) => {
+      sendgridServiceLogger.warn(`Failed to send email on event creation. error=${error}`);
+      return 500;
```
[the remark this answers is not in the document]
Comment (owya490) [reply frontend/services/src/sendgrid/sendgridService.ts:24]: ok i made it raise an error, with the error being dealt with on createEvent and continuing the routing to the new event page as email is not a blocker for creating event.

File: functions/lib/sendgrid/commons.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+PURCHASE_EVENT_EMAIL_TEMPLATE_ID = "d-0c47c68389fe4149a9bf065e38f4a3e9"
+
+def get_user_email(user_id: str, user_data = None) -> str:
+  if user_data == None:
```
[the remark this answers is not in the document]
Comment (owya490) [reply functions/lib/sendgrid/commons.py:7]: done! thanks good catch

**Pull request merged**